openedx_learning/apps/authoring/publishing/models/draft_log.py

```diff
@@ -0,0 +1,237 @@
+    discarding changes (i.e. reset to the published versions) or doing an
+    import. These would be represented by one DraftChangeLog with many
+    DraftChangeLogRecords in it--one DraftChangeLogRecord for every
+    PublishableEntity that was modified.
+
+    Even if we're only directly changing the draft version of one
+    PublishableEntity, we will get multiple DraftChangeLogRecords if changing
+    that entity causes side-effects. See the docstrings for DraftChangeLogRecord
+    and DraftSideEffect for more details.
+    """
+    uuid = immutable_uuid_field()
+    learning_package = models.ForeignKey(LearningPackage, on_delete=models.CASCADE)
+    changed_at = manual_date_time_field()
+    changed_by = models.ForeignKey(
+        settings.AUTH_USER_MODEL,
+        on_delete=models.SET_NULL,
+        null=True,
+        blank=True,
+    )
+
+    class Meta:
+        verbose_name = _("Draft Change Log")
+        verbose_name_plural = _("Draft Change Logs")
+
+
+class DraftChangeLogRecord(models.Model):
+    """
+    A single change in the Draft version of a Publishable Entity.
+
+    We have one unusual convention here, which is that if we have a
+    DraftChangeLogRecord where the old_version == new_version, it means that a
+    Draft's defined version hasn't changed, but the data associated with the
+    Draft has changed because some other entity has changed.
+
+    The main example we have of this are containers. Imagine that we have a
+    Unit that is at version 1, and has unpinned references to various Components
+    that are its children. The Unit's version does not get incremented when the
+    Components are edited, because the Unit container is defined to always get
+    the most recent version of those Components. We would only make a new
+    version of the Unit if we changed the metadata of the Unit itself (e.g. the
+    title), or if we added, removed, or reordered the children.
+
+    Yet updating a Component intuitively changes what we think of as the content
+    of the Unit. Users who are working on Units also expect that a change to a
+    Component will be reflected when looking at a Unit's "last updated" info.
+    The old_version == new_version convention lets us represent that in a useful
+    way because that Unit *is* a part of the change set represented by a
+    DraftChangeLog, even if its own versioned data hasn't changed.
+    """
+    draft_change_log = models.ForeignKey(
+        DraftChangeLog,
+        on_delete=models.CASCADE,
+        related_name="records",
+    )
+    entity = models.ForeignKey(PublishableEntity, on_delete=models.RESTRICT)
+    old_version = models.ForeignKey(
+        PublishableEntityVersion,
+        on_delete=models.RESTRICT,
+        null=True,
+        blank=True,
+        related_name="+",
+    )
+    new_version = models.ForeignKey(
+        PublishableEntityVersion, on_delete=models.RESTRICT, null=True, blank=True
+    )
```

Comment:
Confirming my understanding--let me know if these are wrong. These might be useful as comments on this model.

1. It's valid for multiple DraftChangeLogRecords to exist with the same (entity, old_version, new_version), as long as draft_change_log is distinct for each one. For example, if a user is repeatedly editing a container within a unit U @ v1, we will have a series of DraftChangeLogRecords `(U.v1 -> U.v1), (U.v1 -> U.v1), ... , (U.v1 -> U.v1)`.
2. We cannot assume new_version >= old_version, because discarding changes will be modelled as setting the Draft pointer to an older version--specifically, the last-published version.

Reply:
Both of those statements are correct. I'll add comments for them.

Reply:
I rewrote a lot of the docstring for `DraftChangeLogRecord` in order to illustrate these and other possible scenarios.